### PR TITLE
[WIP] Update a game's ID when advancing to the next scenario.

### LINKED ISSUE #4281: Create new game_info entry for when an MP campaign advances to the next scenario
From Discord/IRC:
```
<Soliton> i meant to mention that mp campaigns are handled poorly so far.
<Soliton> as in advancing to the next scenario.
<Soliton> that should create a new entry.
```

### PR DESCRIPTION
**Pull request by Pentarctagon**
On receiving "store_next_scenario", increment wesnothd's game id counter and assign the game the new id.
Currently this is relevant for when an MP campaign advances to the next scenario, in order to avoid primary key conflicts as well as insert a new game_info row for the new scenario.

Also fixes #4281 by adding a db call in "store_next_scenario" to insert the next scenario's game_info row.

### CODE REVIEW

**Comment by gfgtdf** [comment]
> On receiving "store_next_scenario", increment wesnothd's game id counter and assign the game the new id.

Hmm kinds sounds too easy to me, it seems very likely that there is other code that assumes that the id of a game object does not change during its lifetime if it didnt before

> Also fixes #4281 by adding a db call in "store_next_scenario" to insert the next scenario's game_info row.

thinking about the db, maybe it'D be nice to store the the db that this is a campaign game and what teh previous scenario was? Like having a 'previous_senario_id' in the db.

**Comment by Pentarctagon** [comment]
@gfgtdf Alright, it should be good now.  The second commit, while not strictly related to this PR, was helpful for the below tests I ran and also partially resolves #3635 (to finish that issue I would want to add a button the MP Mod pop up UI as well):

p1,p2,p3 in the lobby
p1,p2 start mp campaign
p3 joins mp campaign as observer
p3 leaves mp campaign
p1,p2 advance to next scenario in mp campaign
p3 joins mp campaign as observer
p2 leaves and then rejoins mp campaign as an observer
p1 sees both p2 and p3 as observers
p2 is given back control of side 2

p1,p2,p3 in the lobby
p1,p2 start mp campaign
p3 uses `/query stopgame p2`

p1,p2,p3 in the lobby
p1,p2 start mp campaign
p1,p2 advance to next scenario in mp campaign
p3 uses `/query stopgame p2`

**Comment by gfgtdf** [comment]
> Alright, it should be good now

I still am not really convinced, it seems to me like `game_id` is used a a key in this multi index so changing it is dangerous.
(similar to usual std::set<>, where changing whether  a > b for element a,b in the set will cause crashes ad random times (the exact time of the crash might depend on the implementation, eg. `std::set<int> s = {1, 2}; *(set.begin()) +=2;` will crash when the set is used later. )

https://github.com/wesnoth/wesnoth/blob/3be0f751cc263065046d44a1f41d2a6131df377f/src/server/player_connection.hpp#L90-L97

so i'm quite sure you have to use the player_connections_.modify() function to make sure the container stays in a valid state.

**Comment by gfgtdf** [comment]
pinging @loonycyborg who wrote this code and might know more.

**Comment by Pentarctagon** [comment]
Would `player_connections_.modify_key()` be preferred, in that case?

**Comment by loonycyborg** [comment]
Yes, all changes to multi-index's elements that change key values must come through modifier functions like modify() or modify_key(). In fact multi-index tend to make it hard to get ahold of non-const reference to an element outside of callbacks for modify(). So if you need to change value of player_record::game_id it needs to be done with one of those funcs.

**Comment by Pentarctagon** [comment]
Alright.  So... how do I actually do that?  I'm pretty well lost on how that's supposed to work.

**Comment by Pentarctagon** [comment]
And, to clarify - `player_record::game_id` isn't a value, it's a function that returns `game_->id()`, which returns the game's `id_` field.  So either way I'm changing the value of `id_`, in which case is the difference that I can't call a function that changes `id_` directly, I have to use modify() to call a function that changes `id_`?

File: src/server/player_connection.cpp [hunk 1]
```diff
@@ -33,6 +33,14 @@ int player_record::game_id() const
 	return game_ ? game_->id() : 0;
 }
 
+void player_record::next_game_id()
```
Comment (soliton-) [inline src/server/player_connection.cpp]: The function name sounds like it's a getter. I'd rather use `set_next_game_id()`.

**Comment by gfgtdf** [comment]
i still think this is not how it works as i understood it you need one `modify` call per player in that game, each one passing that player as the first argument to modify and only. Thinking about it am am in fact unsure whether boost multi index supports 'changing multiple elements simultaniously' at all

**Comment by Pentarctagon** [comment]
It's sounding like it may end up be simpler to just use an auto-increment column as the `game_info` table's primary key.

**Comment by soliton-** [comment]
Another possibility is to actually make it a new game. So if the next scenario data comes in instead of resetting a bunch of things and messing with the multi index and whatnot just create a new game and transfer the players and the few settings that stay the same.

I think in general making it a new game is what one would expect since effectively it really is a new game/scenario. It's certainly that way on the client. It would also fix the possibility of further replays getting the same file name and probably some other edge cases resulting from the ID staying the same.

**Comment by loonycyborg** [comment]
I agree that making a new game would be a cleaner solution.

**Comment by soliton-** [comment]
Btw, if you want to put stopgame into a separate PR I'd merge that right away once help_msg is also updated. :wink:

**Comment by Pentarctagon** [comment]
I'll close this PR, I think, and move stuff around.  Once again I really shouldn't have created a PR from my fork's master branch.

**Comment by loonycyborg** [comment]
> And, to clarify - `player_record::game_id` isn't a value, it's a function that returns `game_->id()`, which returns the game's `id_` field. So either way I'm changing the value of `id_`, in which case is the difference is that I can't call a function that changes `id_` directly, I have to use modify() to call a function that changes `id_`?

Yes. No matter whether it's real data member or calculated value, multi-index must be notified about any possible change of keys by using modify()/modify_key() so it would know to regenerate its internal ordered lists/hash buckets/etc/ . In case of game_id if you change the value of it, you still need to notify multi-index because it keeps a list of references(probably actually binary tree or something) to all elements ordered by value of game_id and it needs to be updated too.

**Pull request closed without merging**